File: app/services/intelligent_retriever.py
"""
Data Vent — Intelligent Retriever

Retrieval engine that routes queries to FalkorDB via vector and graph search.
"""

from __future__ import annotations

import json
from typing import Any, List

import httpx
import structlog
from dataclasses import dataclass, field

from app.services.vector_search import FalkorDBClient

logger = structlog.get_logger(__name__)

@dataclass
class SearchResult:
    """Normalised retrieval result."""
    chunk_id: str
    content: str
    score: float
    metadata: dict[str, Any]
    source: str = "falkordb"
    chunk_type: str = ""
    source_id: str = ""
    document_id: str = ""
    depth: int = 0

    def to_dict(self) -> dict[str, Any]:
        return {
            "chunk_id": self.chunk_id,
            "content": self.content,
            "score": self.score,
            "metadata": self.metadata,
            "source": self.source,
            "chunk_type": self.chunk_type,
            "source_id": self.source_id,
            "document_id": self.document_id,
            "depth": self.depth,
        }

@dataclass
class DFSTraversalResult:
    chunks: List[SearchResult]

class IntelligentRetriever:
    """
    Retrieval engine using FalkorDBClient and embeddings service.
    """

    def __init__(self, falkordb_client: FalkorDBClient, settings: Any) -> None:
        self._falkordb = falkordb_client
        self.ollama_url = getattr(settings, "OLLAMA_URL", "http://localhost:11434")
        self._http_client = httpx.AsyncClient(timeout=15.0)

    @property
    def active_backend(self) -> str:
        """Return the name of the currently active retrieval backend."""
        return "falkordb"

    async def close(self) -> None:
        await self._http_client.aclose()

    async def vectorize_query(self, query: str) -> List[float]:
        """Generate embeddings for a single query string using Ollama directly."""
        try:
            response = await self._http_client.post(
                f"{self.ollama_url}/api/embed",
                json={"input": [query], "model": "nomic-embed-text"}
            )
            response.raise_for_status()
            data = response.json()
            embeddings = data.get("embeddings", [])
            return embeddings[0] if embeddings else []
        except Exception as e:
            logger.error("vectorize_query_failed", query=query[:80], error=str(e))
            return []

    async def vector_search(self, query_vectors: List[float], limit: int = 10) -> List[SearchResult]:
        """Perform vector similarity search on FalkorDB."""
        if not query_vectors:
            return []
            
        try:
            embedding_str = json.dumps(query_vectors)
            cypher = f"""
            CALL db.idx.vector.queryNodes('Vector_Chunk', 'embeddings', {limit}, vecf32({embedding_str})) YIELD node, score
            RETURN node.id AS chunk_id,
                   node.content AS content,
                   node.chunk_type AS chunk_type,
                   node.source_id AS source_id,
                   node.metadata AS metadata,
                   score AS score
            ORDER BY score DESC
            """
            raw_result = await self._falkordb.query(cypher)
            return self._parse_graph_results(raw_result, is_vector=True)
        except Exception as e:
            logger.error("vector_search_failed", error=str(e))
            return []

    async def text_search(self, query: str, limit: int = 10) -> List[SearchResult]:
        """Perform Full-Text Search on FalkorDB as a fallback."""
        try:
            import re
            words = re.findall(r'\b\w+\b', query.lower())
            keywords = [w for w in words if len(w) > 2]
            if not keywords and words:
                keywords = words
                
            if not keywords:
                return []
                
            fts_query = " ".join(keywords)
            
            cypher = f"""
            CALL db.idx.fulltext.queryNodes('Vector_Chunk', '{fts_query}') YIELD node, score
            RETURN node.id AS chunk_id,
                   node.content AS content,
                   node.chunk_type AS chunk_type,
                   node.source_id AS source_id,
                   node.metadata AS metadata,
                   score AS score
            ORDER BY score DESC
            LIMIT {limit}
            """
            raw_result = await self._falkordb.query(cypher)
            return self._parse_graph_results(raw_result, is_vector=True)
        except Exception as e:
            logger.error("text_search_failed", error=str(e))
            return []

    async def dfs_traversal(
        self,
        start_chunk_ids: List[str],
        max_depth: int = 2,
        min_relevance: float = 0.3,
        max_results: int = 20,
    ) -> DFSTraversalResult:
        """Perform variable-length path traversal from start nodes in FalkorDB."""
        if not start_chunk_ids:
            return DFSTraversalResult(chunks=[])
            
        try:
            # We construct a Cypher query using an IN clause for the start nodes
            ids_str = ", ".join(f"'{cid}'" for cid in start_chunk_ids)
            
            # Using variable length path. We cap the depth.
            cypher = f"""
            MATCH path = (start:Vector_Chunk)-[*1..{max_depth}]-(n:Vector_Chunk)
            WHERE start.id IN [{ids_str}] AND NOT n.id IN [{ids_str}]
            RETURN DISTINCT n.id AS chunk_id,
                   n.content AS content,
                   n.chunk_type AS chunk_type,
                   n.source_id AS source_id,
                   n.metadata AS metadata,
                   length(path) AS depth
            LIMIT {max_results}
            """
            raw_result = await self._falkordb.query(cypher)
            chunks = self._parse_graph_results(raw_result, is_vector=False)
            
            # Base graph score decreases with depth.
            for chunk in chunks:
                chunk.score = max(0.1, 1.0 - (chunk.depth * 0.2))
                
            # Filter by relevance
            chunks = [c for c in chunks if c.score >= min_relevance]
            
            return DFSTraversalResult(chunks=chunks)
            
        except Exception as e:
            logger.error("dfs_traversal_failed", error=str(e))
            return DFSTraversalResult(chunks=[])

    def _parse_graph_results(self, raw_result: Any, is_vector: bool) -> List[SearchResult]:
        """Parse raw GRAPH.QUERY output into SearchResult objects."""
        # FalkorDB typically returns [[headers], [[row1], [row2]], [stats]]
        if not raw_result or len(raw_result) < 2:
            return []
            
        headers_raw = raw_result[0]
        rows_raw = raw_result[1]
        
        # Decode headers
        headers = []
        for h in headers_raw:
            if isinstance(h, bytes):
                headers.append(h.decode('utf-8'))
            else:
                headers.append(str(h))
                
        results = []
        for row in rows_raw:
            chunk_id = ""
            content = ""
            chunk_type = ""
            source_id = ""
            metadata_str = "{}"
            score = 0.0
            depth = 0
            
            for i, val in enumerate(row):
                if i >= len(headers):
                    continue
                    
                col_name = headers[i]
                
                # Extract value
                v = val
                if isinstance(val, bytes):
                    v = val.decode('utf-8')
                    
                if col_name == "chunk_id":
                    chunk_id = str(v)
                elif col_name == "content":
                    content = str(v)
                elif col_name == "chunk_type":
                    chunk_type = str(v)
                elif col_name == "source_id":
                    source_id = str(v)
                elif col_name == "metadata":
                    metadata_str = str(v) if v else "{}"
                elif col_name == "score" and is_vector:
                    try:
                        score = float(v)
                    except (ValueError, TypeError):
                        pass
                elif col_name == "depth":
                    try:
                        depth = int(v)
                    except (ValueError, TypeError):
                        pass
                        
            metadata = {}
            try:
                if metadata_str and metadata_str != "None":
                    metadata = json.loads(metadata_str)
            except json.JSONDecodeError:
                pass
                
            results.append(SearchResult(
                chunk_id=chunk_id,
                content=content,
                score=score,
                metadata=metadata,
                chunk_type=chunk_type,
                source_id=source_id,
                document_id=source_id,
                depth=depth,
            ))
            
        return results

    async def retrieve(
        self,
        query: str,
        group_ids: list[str] | None = None,
        num_results: int = 10,
        center_node_uuid: str | None = None,
    ) -> list[SearchResult]:
        vector = await self.vectorize_query(query)
        results = []
        if vector:
            results = await self.vector_search(vector, num_results)
            
        # Fallback to text search if vector search returns 0 results (e.g. embeddings service was down)
        if not results:
            logger.info("Vector search returned 0 results, falling back to text search", query=query)
            results = await self.text_search(query, num_results)
            
        return results

    async def retrieve_with_context(
        self,
        query: str,
        entity_uuid: str,
        group_ids: list[str] | None = None,
        num_results: int = 15,
    ) -> list[SearchResult]:
        return await self.retrieve(query, group_ids, num_results, center_node_uuid=entity_uuid)

File: src/services/intelligent_retriever.rs
use serde::{Deserialize, Serialize};
use tracing::{error, info, warn};

use crate::config::Config;
use crate::services::vector_search::FalkorDBClient;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult {
    pub chunk_id: String,
    pub content: String,
    pub score: f64,
    pub metadata: serde_json::Value,
    pub source: String,
    pub chunk_type: String,
    pub source_id: String,
    pub document_id: String,
    pub depth: i32,
    #[serde(default)]
    pub matched_by_chunks: Vec<String>,
}

pub struct IntelligentRetriever {
    falkordb_client: FalkorDBClient,
    http_client: reqwest::Client,
    gemini_api_key: String,
    gemini_base_url: String,
    embedding_model: String,
}

impl IntelligentRetriever {
    pub fn new(falkordb_client: FalkorDBClient, config: &Config) -> Self {
        Self {
            falkordb_client,
            http_client: reqwest::Client::builder()
                .timeout(std::time::Duration::from_secs(15))
                .build()
                .unwrap_or_default(),
            gemini_api_key: config.gemini_api_key.clone(),
            gemini_base_url: config.gemini_base_url.clone(),
            embedding_model: config.gemini_embedding_model.clone(),
        }
    }

    pub async fn _close(&self) {
        let _ = self.falkordb_client._close().await;
    }

    pub async fn vectorize_query(&self, query: &str) -> Vec<f64> {
        if self.gemini_api_key.is_empty() {
            warn!("GEMINI_API_KEY not set");
            return vec![];
        }

        let url = format!(
            "{}/v1/models/{}:embedContent?key={}",
            self.gemini_base_url, self.embedding_model, self.gemini_api_key
        );

        let body = serde_json::json!({
            "model": format!("models/{}", self.embedding_model),
            "content": {
                "parts": [{"text": query}]
            }
        });

        match self.http_client.post(&url).json(&body).send().await {
            Ok(resp) => {
                if let Ok(data) = resp.json::<serde_json::Value>().await {
                    if let Some(embedding) = data.get("embedding") {
                        if let Some(values) = embedding.get("values") {
                            if let Some(arr) = values.as_array() {
                                return arr.iter().filter_map(|v| v.as_f64()).collect();
                            }
                        }
                    }
                }
            }
            Err(e) => {
                error!("vectorize_query_failed: {}", e);
            }
        }
        vec![]
    }

    pub async fn vector_search(&self, graph_name: &str, query_vectors: &[f64], limit: usize) -> Vec<SearchResult> {
        if query_vectors.is_empty() {
            return vec![];
        }

        let embedding_str = serde_json::to_string(query_vectors).unwrap_or_default();
        let cypher = format!(
            "CALL db.idx.vector.queryNodes('Vector_Chunk', 'embeddings', {}, vecf32({})) YIELD node, score \
             RETURN node.id AS chunk_id, node.content AS content, node.chunk_type AS chunk_type, \
             node.source_id AS source_id, node.metadata AS metadata, score AS score \
             ORDER BY score DESC",
            limit, embedding_str
        );

        match self.falkordb_client.query(graph_name, &cypher).await {
            Ok(val) => self.parse_graph_results(&val, true),
            Err(e) => {
                error!("vector_search_failed: {}", e);
                vec![]
            }
        }
    }

    pub async fn text_search(&self, graph_name: &str, query: &str, limit: usize) -> Vec<SearchResult> {
        // Implement simple text search via regex for word boundaries and length > 2
        let words: Vec<&str> = query.split_whitespace().collect();
        let keywords: Vec<&str> = words.into_iter().filter(|w| w.len() > 2).collect();
        if keywords.is_empty() {
            return vec![];
        }
        let fts_query = keywords.join(" ");
        let cypher = format!(
            "CALL db.idx.fulltext.queryNodes('Vector_Chunk', '{}') YIELD node, score \
             RETURN node.id AS chunk_id, node.content AS content, node.chunk_type AS chunk_type, \
             node.source_id AS source_id, node.metadata AS metadata, score AS score \
             ORDER BY score DESC LIMIT {}",
            fts_query, limit
        );

        match self.falkordb_client.query(graph_name, &cypher).await {
            Ok(val) => self.parse_graph_results(&val, true),
            Err(e) => {
                error!("text_search_failed: {}", e);
                vec![]
            }
        }
    }

    pub async fn dfs_traversal(
        &self,
        graph_name: &str,
        start_chunk_ids: &[String],
        max_depth: usize,
        min_relevance: f64,
        max_results: usize,
    ) -> Vec<SearchResult> {
        if start_chunk_ids.is_empty() {
            return vec![];
        }

        let ids_str = start_chunk_ids
            .iter()
            .map(|id| format!("'{}'", id))
            .collect::<Vec<_>>()
            .join(", ");

        let cypher = format!(
            "MATCH path = (start:Vector_Chunk)-[*1..{}]-(n:Vector_Chunk) \
             WHERE start.id IN [{}] AND NOT n.id IN [{}] \
             RETURN DISTINCT n.id AS chunk_id, n.content AS content, n.chunk_type AS chunk_type, \
             n.source_id AS source_id, n.metadata AS metadata, length(path) AS depth \
             LIMIT {}",
            max_depth, ids_str, ids_str, max_results
        );

        match self.falkordb_client.query(graph_name, &cypher).await {
            Ok(val) => {
                let mut chunks = self.parse_graph_results(&val, false);
                for chunk in &mut chunks {
                    chunk.score = 0.1f64.max(1.0 - (chunk.depth as f64 * 0.2));
                }
                chunks.into_iter().filter(|c| c.score >= min_relevance).collect()
            }
            Err(e) => {
                error!("dfs_traversal_failed: {}", e);
                vec![]
            }
        }
    }

    fn parse_graph_results(&self, raw_result: &serde_json::Value, is_vector: bool) -> Vec<SearchResult> {
        let mut results = vec![];
        if let Some(arr) = raw_result.as_array() {
            if arr.len() >= 2 {
                let headers_raw = &arr[0];
                let rows_raw = &arr[1];

                let mut headers = vec![];
                if let Some(h_arr) = headers_raw.as_array() {
                    for h in h_arr {
                        if let Some(s) = h.as_str() {
                            headers.push(s.to_string());
                        } else {
                            headers.push(h.to_string());
                        }
                    }
                }

                if let Some(r_arr) = rows_raw.as_array() {
                    for row in r_arr {
                        if let Some(cols) = row.as_array() {
                            let mut chunk_id = String::new();
                            let mut content = String::new();
                            let mut chunk_type = String::new();
                            let mut source_id = String::new();
                            let mut metadata = serde_json::Value::Null;
                            let mut score = 0.0;
                            let mut depth = 0;

                            for (i, val) in cols.iter().enumerate() {
                                if i >= headers.len() {
                                    continue;
                                }
                                let col_name = &headers[i];
                                let v_str = val.as_str().unwrap_or("").to_string();

                                match col_name.as_str() {
                                    "chunk_id" => chunk_id = v_str,
                                    "content" => content = v_str,
                                    "chunk_type" => chunk_type = v_str,
                                    "source_id" => source_id = v_str,
                                    "metadata" => {
                                        if !v_str.is_empty() && v_str != "None" {
                                            if let Ok(m) = serde_json::from_str(&v_str) {
                                                metadata = m;
                                            }
                                        }
                                    }
                                    "score" if is_vector => {
                                        if let Some(n) = val.as_f64() {
                                            score = n;
                                        } else if let Ok(n) = v_str.parse::<f64>() {
                                            score = n;
                                        }
                                    }
                                    "depth" => {
                                        if let Some(n) = val.as_i64() {
                                            depth = n as i32;
                                        } else if let Ok(n) = v_str.parse::<i32>() {
                                            depth = n;
                                        }
                                    }
                                    _ => {}
                                }
                            }

                            results.push(SearchResult {
                                chunk_id,
                                content,
                                score,
                                metadata,
                                source: "falkordb".to_string(),
                                chunk_type,
                                document_id: source_id.clone(),
                                source_id,
                                depth,
                                matched_by_chunks: vec![],
                            });
                        }
                    }
                }
            }
        }
        results
    }

    pub async fn retrieve(
        &self,
        graph_name: &str,
        query: &str,
        _group_ids: Option<Vec<String>>,
        num_results: usize,
        _center_node_uuid: Option<String>,
    ) -> Vec<SearchResult> {
        let vector = self.vectorize_query(query).await;
        let mut results = vec![];

        if !vector.is_empty() {
            results = self.vector_search(graph_name, &vector, num_results).await;
        }

        if results.is_empty() {
            info!("Vector search returned 0 results, falling back to text search for: {}", query);
            results = self.text_search(graph_name, query, num_results).await;
        }
        results
    }
}
